src/libopensc/card-myeid.c

```diff
@@ -1185,6 +1261,149 @@ static int myeid_decipher(struct sc_card *card, const u8 * crgram,
+	u8 p2 = 0x86; /* init P2 for asymmetric crypto by default.*/
+	myeid_private_data_t* priv;
+	int symmetric_operation = 0;
+
+	LOG_FUNC_CALLED(card->ctx);
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_NORMAL);
+
+	if (card == NULL || crgram == NULL) {
+		LOG_TEST_RET(card->ctx, SC_ERROR_INVALID_ARGUMENTS, "One or more of required arguments was null.\n");
+	}
+
+	if (crgram_len > MYEID_MAX_RSA_KEY_LEN / 8)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+
+	if (card->drv_data)
+	{
+		priv = (myeid_private_data_t*) card->drv_data;
+
+		if (priv->sec_env)
+		{
+			if (priv->sec_env->algorithm == SC_ALGORITHM_AES ||
+				priv->sec_env->algorithm == SC_ALGORITHM_3DES ||
+				priv->sec_env->algorithm == SC_ALGORITHM_DES)
+					symmetric_operation = 1;
+		}
+	}
+
+	if (symmetric_operation)
+	{
+		p2 = 0x84;		/* Set correct P2 for symmetric crypto */
+		if (crgram_len > MYEID_MAX_APDU_DATA_LEN)
+		{
+			LOG_TEST_RET(card->ctx, SC_ERROR_WRONG_LENGTH, "Unwrapping symmetric data longer that 255 bytes is not supported\n");
+		}
+	}
+
+	/* INS: 0x2A  PERFORM SECURITY OPERATION
+	    * P1:  0x00  Do not expect response - the deciphered data will be placed into the target key EF.
+	    * P2:  0x86  Cmd: Padding indicator byte followed by cryptogram
+		* P2:  0x84  Cmd: AES/3DES Cryptogram (plain value encoded in BER-TLV DO, but not including SM DOs)
+		If crgram_len == 256 (2048 bit RSA), we split the cryptogram in two and send two APDUs.
+	 */
+	sc_format_apdu(card, &apdu,
+		SC_APDU_CASE_3_SHORT,
+		0x2A, 0x00, p2);
+
+	apdu.resp = rbuf;
+	apdu.resplen = sizeof(rbuf);
+	apdu.le = 0;
+
+	if (symmetric_operation)
+	{
+		/* symmetric crypto, no padding indicator byte */
+		{
+			memcpy(sbuf, crgram, crgram_len);
+			apdu.lc = crgram_len;
+		}
+	}
+	else
+	{
+		if (crgram_len > MYEID_MAX_APDU_DATA_LEN)
+		{
+			/* padding indicator byte, 0x81 = first half of 2048 bit cryptogram */
+			sbuf[0] = 0x81;
+			memcpy(sbuf + 1, crgram, crgram_len / 2);
+			apdu.lc = crgram_len / 2 + 1;
+		}
+		else
+		{
+			sbuf[0] = 0; /* padding indicator byte, 0x00 = No further indication */
+			memcpy(sbuf + 1, crgram, crgram_len);
+			apdu.lc = crgram_len + 1;
+		}
+	}
+
+	apdu.datalen = apdu.lc;
+	apdu.data = sbuf;
+
+	r = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
+	if (apdu.sw1 == 0x90 && apdu.sw2 == 0x00)
+	{
+		if (crgram_len > MYEID_MAX_APDU_DATA_LEN)
+		{
+			sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT,
+				0x2A, 0x00, p2);
+			apdu.resp = rbuf;
+			apdu.resplen = sizeof(rbuf);
+			apdu.le = 0;
+			/* padding indicator byte,
+			 * 0x82 = Second half of 2048 bit cryptogram */
+			sbuf[0] = 0x82;
+			memcpy(sbuf + 1, crgram + crgram_len / 2, crgram_len / 2);
+			apdu.lc = crgram_len / 2 + 1;
+			apdu.datalen = apdu.lc;
+			apdu.data = sbuf;
+
+			r = sc_transmit_apdu(card, &apdu);
+
+			LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
+		}
+	}
+	LOG_FUNC_RETURN(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2));
+}
```

Comment:
In line 1312 we validate crgram_len with this check, which is supposed to prevent buffer overrun in the later memcpys :

	if (crgram_len > MYEID_MAX_RSA_KEY_LEN / 8)
	     LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);

MYEID_MAX_RSA_KEY_LEN / 8 = 256.

For symmetric keys we have a separate check.

This already prevents buffer overrun in the later memcpys. When we increase MYEID_MAX_RSA_KEY_LEN, we have to work with all functions that do RSA crypto anyway, to support longer key sizes. So I think there's not much risk that we would change this constant and forget to update the argument validation at that point.

Well, actually I can find one situation which could go wrong… If crgram_len would be 255, there's a risk of buffer overrun here:

		sbuf[0] = 0; /* padding indicator byte, 0x00 = No further indication */
			memcpy(sbuf + 1, crgram, crgram_len);

By changing:

if (crgram_len > MYEID_MAX_APDU_DATA_LEN)
to
if (crgram_len >= MYEID_MAX_APDU_DATA_LEN)

are we on the safe side?

Can you find some other specific situation which could cause buffer overrun?

Reply:
After all even the issue above wouldn't cause a buffer overrun. sbuf is SC_MAX_APDU_BUFFER_SIZE long which is 261 bytes so there is a margin already. But is there something else?

Reply:
You're right it wouldn't cause any problems, currently. But it would be more safe and more obvious if you're checking the length when it's actually used. Hence, I would not compare against `MYEID_MAX_APDU_DATA_LEN` and instead compare `crgram_len` with `sizeof sbuf`...

Reply:
many card drivers overuse the SC_MAX_APDU_BUFFER_SIZE and get away with it. 
it is defined in types.h as:
```
#define SC_MAX_APDU_BUFFER_SIZE   261 /* takes account of: CLA INS P1 P2 Lc [255 byte of data] Le */
```
It defines the max size of a short APDU with data to be sent to the reader in a single buffer.

The code is also being written with the max RSA key size of 2048. 
May I suggest that you make the code more general in order to handle large RSA keys when they come out with a new card that supports a 4096 byte RSA key?

In other words base you buffer sizes of data to be sent or received on MYEID_MAX_RSA_KEY_LEN Then only deal with having to send it in chucks with the extra leading flag byte within a loop rather then the if at 1361. 

You divide the data in half to send in two equal size chunks. Is that a card restriction? Most drivers send as much as possible in an APDU (card->max_send_size), then the rest in the final APDU. This then can work with readers that may support max_send_size < 255.

Reply:
Currently we only need to split the APDU in two if the key size is 2048 bits so the logic is fixed as it is. Our implementation is to split it in half. We are going to release a new MyEID version with 4096 bit keys hopefully in the near future. We are already working with it. 

When we proceed into implementing support for 4096 bit keys into OpenSC we have to rewrite these functions (unwrap, deciper, compute_signature) anyway. It's better to use a loop then and make it more generic. As long as we can agree that the function is safe as it is, I'd prefer to not put effort in refactoring it now to avoid double work on the same issue. When we have the 4K keys, I'd rather improve the code in all of these functions at the same time and create a separate PR for this. 

I hope we could complete merging this PR asap because some of the code is already quite old and rebasing it is time consuming if same parts of code will be changed again in the main branch. I'll provide some more test results soon.